freqtrade/freqai/utils.py

```diff
@@ -0,0 +1,138 @@
+from freqtrade.exchange import timeframe_to_seconds
+from freqtrade.exchange.exchange import market_is_active
+from freqtrade.plugins.pairlist.pairlist_helpers import dynamic_expand_pairlist
+
+
+logger = logging.getLogger(__name__)
+
+
+def download_all_data_for_training(dp: DataProvider, config: dict) -> None:
+    """
+    Called only once upon start of bot to download the necessary data for
+    populating indicators and training the model.
+    :param timerange: TimeRange = The full data timerange for populating the indicators
+                                    and training the model.
+    :param dp: DataProvider instance attached to the strategy
+    """
+
+    if dp._exchange is not None:
+        markets = [p for p, m in dp._exchange.markets.items() if market_is_active(m)
+                   or config.get('include_inactive')]
+    else:
+        # This should not occur:
+        raise OperationalException('No exchange object found.')
+
+    all_pairs = dynamic_expand_pairlist(config, markets)
+
+    timerange = get_required_data_timerange(config)
+
+    new_pairs_days = int((timerange.stopts - timerange.startts) / 86400)
+    if not dp._exchange:
+        # Not realistic - this is only called in live mode.
+        raise OperationalException("Dataprovider did not have an exchange attached.")
+    refresh_backtest_ohlcv_data(
+        dp._exchange,
+        pairs=all_pairs,
+        timeframes=config["freqai"]["feature_parameters"].get("include_timeframes"),
+        datadir=config["datadir"],
+        timerange=timerange,
+        new_pairs_days=new_pairs_days,
+        erase=False,
+        data_format=config.get("dataformat_ohlcv", "json"),
+        trading_mode=config.get("trading_mode", "spot"),
+        prepend=config.get("prepend_data", False),
+    )
+
+
+def get_required_data_timerange(
+    config: dict
+) -> TimeRange:
+    """
+    Used to compute the required data download time range
+    for auto data-download in FreqAI
+    """
+    time = datetime.now(tz=timezone.utc).timestamp()
+    data_load_timerange = TimeRange()
+
+    timeframes = config["freqai"]["feature_parameters"].get("include_timeframes")
+
+    max_tf_seconds = 0
+    for tf in timeframes:
+        secs = timeframe_to_seconds(tf)
+        if secs > max_tf_seconds:
+            max_tf_seconds = secs
+
+    startup_candles = config.get('startup_candle_count', 0)
+    indicator_periods = config["freqai"]["feature_parameters"]["indicator_periods_candles"]
+
+    # factor the max_period as a factor of safety.
+    max_period = int(max(startup_candles, max(indicator_periods)) * 1.5)
+    config['startup_candle_count'] = max_period
+    logger.info(f'FreqAI auto-downloader using {max_period} startup candles.')
+
+    additional_seconds = max_period * max_tf_seconds
+
+    data_load_timerange.startts = int(
+        time
+        - config["freqai"].get("train_period_days", 0) * 86400
+        - additional_seconds
+    )
+    data_load_timerange.stopts = int(time)
```

Comment:
You can't really (reliably) assign a timeRange object like this.

wherever we use this - we're checking for `timerange.starttype == 'date'` (or at least we should) - and by not assigning that, it might mean that it'll not use the timerange given.

You can instead use `Timerange('date', 'date', startts, stopts)` to create the instance once the values are clear (instead of creating an empty instance and then reassigning attributes).